src/components/MessagesList/MessagesList.vue

```diff
@@ -406,49 +406,38 @@ export default {
 		},
 
 		softUpdateAuthorGroups(oldGroups, newGroups, dateTimestamp) {
-			const newGroupIdSet = new Set(this.messagesList.map(message => message.id))
-			const newGroupsKeys = Object.keys(newGroups)
+			const groupIds = new Set([...Object.keys(oldGroups), ...Object.keys(newGroups)])
 
-			Object.entries(oldGroups).forEach(([id, oldGroup]) => {
-				if (newGroupsKeys.includes(id)) {
-					// Group is presented, will be handled in the next loop
-					return
-				}
-
-				if (oldGroup.messages.some(message => !newGroupIdSet.has(message.id))) {
-					// Delete groups of normal and temporary messages,
-					// if at least one message from the group is no longer in the store
+			groupIds.forEach(id => {
+				if (oldGroups[id] && !newGroups[id]) {
+					// group no longer exists, remove
 					this.$delete(this.messagesGroupedByDateByAuthor[dateTimestamp], id)
-				}
-			})
-			Object.entries(newGroups).forEach(([id, newGroup]) => {
-				if (!oldGroups[id]) {
-					const oldId = Object.keys(oldGroups)
-						.find(key => +id < +key && oldGroups[key].nextMessageId <= newGroup.nextMessageId)
-					if (oldId) {
-						// newGroup includes oldGroup and more old messages, remove oldGroup
-						this.$delete(this.messagesGroupedByDateByAuthor[dateTimestamp], oldId)
-					}
-					// newGroup is not presented in the list, add it
-					this.$set(this.messagesGroupedByDateByAuthor[dateTimestamp], id, newGroup)
-				} else if (!this.areGroupsIdentical(newGroup, oldGroups[id])) {
-					// newGroup includes oldGroup and more recent messages
-					this.$set(this.messagesGroupedByDateByAuthor[dateTimestamp], id, newGroup)
+				} else if (newGroups[id] && !oldGroups[id]) {
+					// group did not exist before, add
+					this.$set(this.messagesGroupedByDateByAuthor[dateTimestamp], id, newGroups[id])
+				} else if (!this.areGroupsIdentical(newGroups[id], oldGroups[id])) {
+					// group exists in both, update
+					this.$set(this.messagesGroupedByDateByAuthor[dateTimestamp], id, newGroups[id])
 				}
 			})
 		},
 
 		areGroupsIdentical(group1, group2) {
+			// Compare plain values
 			if (group1.messages.length !== group2.messages.length
-				|| JSON.stringify(group1.messages) !== JSON.stringify(group2.messages)
 				|| group1.dateSeparator !== group2.dateSeparator
 				|| group1.previousMessageId !== group2.previousMessageId
 				|| group1.nextMessageId !== group2.nextMessageId) {
 				return false
 			}
 
-			// Check for temporary messages, replaced with messages from server
-			return group1.messages.every((message, index) => group2.messages[index].id === message.id)
+			// Compare ids (temporary messages, replaced with messages from server)
+			if (!group1.messages.every((message, index) => group2.messages[index].id === message.id)) {
+				return false
+			}
+
+			// Compare stringified arrays of messages
+			return JSON.stringify(group1.messages) === JSON.stringify(group2.messages)
```

Comment:
Could `JSON.stringify` comparison be enough to also spot if there are some temp messages replaced from server ( ids are not the same so it will return false)?

Reply:
it would be enough. But:
- JSON conversion is complex and requires expensive computations
- after conversion, we need to compare two strings character by character (minimum converted string length for single message is ~400 characters)
- with id comparison, we check only N pairs of values - make sense to try it first

Reply:
Yes but temp messages are usually  in 1 group and JSON conversion would thus be applied on n-1 items while 1 group is exited with id comparison.
The question is if n-1 id comparison pairs are cheaper than 1 JSON coversion/comparison.